documentation/writing_tests.md

```diff
@@ -0,0 +1,1055 @@
+- [Recommended tools](#recommended-tools)
+    - [Mocha](#mocha)
+    - [Chai](#chai)
+    - [Rollup](#rollup)
+    - [Karma](#karma)
+    - [Recorder](#recorder)
+- [Test folder structure](#test-folder-structure)
+    - [Testing cloud resources](#testing-cloud-resources)
+    - [Public or internal tests](#public-or-internal-tests)
+    - [Testing API functionalities](#testing-api-functionalities)
+- [Shared and reusable code](#shared-and-reusable-code)
+    - [Preparing all of the test cases](#preparing-all-of-the-test-cases)
+    - [Preparing some of the test cases](#preparing-some-of-the-test-cases)
+    - [Universal utilities](#universal-utilities)
+- [Writing test cases](#writing-test-cases)
+    - [What a test is actually testing](#what-a-test-is-actually-testing)
+    - [Test titles](#test-titles)
+    - [Inner parts of a test](#inner-parts-of-a-test)
+    - [Using conditionals](#using-conditionals)
+    - [Using delays](#using-delays)
+    - [Exceptions and edge cases](#exceptions-and-edge-cases)
+    - [How tests should look](#how-tests-should-look)
+- [Getting feedback](#getting-feedback)
+
+
+
+## Engineering setup
+
+The Azure SDK tests are valuable due to many factors. They work as a way to verify that our code is correct, just as much as a way to share how to use our code with our customers. To monitor that our tests are working correctly, they are triggered by automatic systems that help us verify that our commits are correct, or check that the services we're targeting don't show unexpected behaviors, all of which help us have a better level of confidence before releasing anything to the public. 
+
+For our Engineering Systems to pick up our tests appropriately, our packages must be configured according to their guidelines. In this section we will go through some of these concepts, and provide links that expand them in detail. We will be covering:
+
+- [Engineering goals](#engineering-goals).
+- [CI and nightly test configuration](#ci-and-nightly-test-configuration).
+- [Delivering live tests to our users](#delivering-live-tests-to-our-users).
+
+### Engineering goals
+
+Though the tests for the Azure SDK for JavaScript and TypeScript must target live resources, we should make sure they only do so when necessary. For this purpose, we must keep in mind the following guidelines:
+
+- Tests should not be flaky. Tests should pass regardless of who's executing it, when they are running, and how many times they run.
+- Tests should create the resources they are testing.
+- While writing tests, use your own personal resources for setting up the context in which each test will create their own resources. For example, it is valid to have a static KeyVault while writing KeyVault tests, and then a given test can create a KeyVault Key before validating any of the functionalities of the KeyVault Key. Ask your team to see if there's a resource already in place for test development.
```

Comment:
Do we have any recommended tools to help developers avoid committing secrets?

Reply:
What kind of tool do you mean? We have plenty of tools related to credentials for tests, but I want to understand what you're looking for.

Reply:
I was wondering if there was some tool that can alert the developer that a password or a service key is about to be committed? 

However, since we're recommending the use of environment variables everywhere this is probably not very important

Reply:
I'm guessing it's going to be `dotenv`, which I think we standardized on all the way into our samples.

Reply:
GitHub is supposed to have some support for cred scanning and I think we can add a CI gate for it? It wouldn't help you not expose the cred (since you already pushed) but it would at least tell you to rotate it.

Reply:
but yes realistically I like always using env vars for this